### ISSUE TITLE
RFC: Private CUDA memory pools

### ISSUE THREAD

**Issue by mcarilli** [pytorch/pytorch#51075]
## 🚀 Feature

Allow users to specify regions where CUDA memory allocations are satisfied from a region-private pool.

## Motivation

CUDA graph capture is our main motivation. But it seems like a handy thing, there may be other uses I don't anticipate.

[CUDA graph capture](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#creating-a-graph-using-stream-capture) performs a dry run of a region of execution, freezing all CUDA work (and virtual addresses used during that work) into a "graph." The graph may be "replayed" like a single giant kernel, with greatly reduced CPU overhead as well as modestly improved GPU performance.

Because capture bakes in memory addresses, the memory used during capture must be available for the graph to use during replay. But Pytorch's current caching allocator assigns and frees memory eagerly and dynamically, so when a graph is replayed, those memory addresses may be in use by other tensors. One way to guarantee a graph's baked in addresses are always safe to reuse is to satisfy allocation requests from a graph-private memory pool during capture.

## Pitch

The simplest API that comes to mind is something like
```python
pool = torch.cuda.MemPool() # MemPool would be a simple Python object, its only data member would be an integer uuid.
with torch.cuda.mempool(pool):
    # all tensors created in this region have their allocations satisfied from the private pool.
    # Capture a graph here
    # Within capture, temporary internal reallocations can be assigned, freed, and reassigned as usual, as long as the high-water
    # mark of memory blocks used during capture survives and isn't in use by any other tensors when the graph is later replayed.
    # This will hold as long as the pool is private and no cudaFrees occur in the pool between now and replay().
```

The simplest implementation that comes to mind (serving the above API) would be a per-pool-id list of [`THCCachingAllocator`](https://github.com/pytorch/pytorch/blob/a347c747df8302acc0007a26f23ecf3355a5bef9/c10/cuda/CUDACachingAllocator.cpp#L945)s (or [`DeviceCachingAllocator`](https://github.com/pytorch/pytorch/blob/a347c747df8302acc0007a26f23ecf3355a5bef9/c10/cuda/CUDACachingAllocator.cpp#L183)s).

To make graph replay safer, private (non-default) memory pools could optionally be told to error on any internal calls to cudaFree (which might mistakenly free addresses out from under graphs previously captured using that pool):
```
pool = torch.cuda.MemPool(error_on_free=True)
```

@arslan-zulfiqar made some modifications to the current allocator in a local fork to implement region-private memory pools. We've used his build to run graph-captured BERT and Mask-RCNN, and it works, including for complex cases like running some segments of the model as graphed and some segments eagerly (which is essential to if some segments are uncapturable, for example, if they contain data dependent control flow). So we've demonstrated enabling region-specific private pools can make it safe to capture the current allocator.

#### 1 (API alternative): reduce API surface by having graphs request the pool
This is an API alternative, not an implementation alternative. Initiating region-private allocation behavior could be folded into `torch.cuda._Graph.capture_begin()`, so the user wouldn't need a `with private_mempool`.

#### 2 (API and implementation alternative): ability to request a unique stream
This is both an API and implementation alternative.

The current allocator silos allocations per-stream. So running a graph-capture on a side stream (which the capture API requires anyway) effectively gives you a private pool, as long as you can be sure no one else uses the stream. Right now you don't have that certainty: Pytorch uses a pool of streams under the hood, so there's no guarantee your side stream won't alias another side stream requested elsewhere in the script. However, it wouldn't be hard to expose an API that constructs and returns a unique (not from the stream pool) stream that no other stream will alias. Then you could say
```python
stream = torch.cuda.Stream(unique=True)
with torch.cuda.stream(stream):
    # capture a graph here, memory is siloed for you
```

#### 3 (Implementation alternative): Silo allocations by pool id within one Allocator object

Instead of a list of per-pool-id THCCachingAllocators (or DeviceCachingAllocators), we could silo allocations per-pool-id by minimally extending the current stream siloing logic. Specifically, we could add
```
if (a->pool_id != b->pool_id) {
  return a->pool_id < b->pool_id;
}
```
[here](https://github.com/pytorch/pytorch/blob/a347c747df8302acc0007a26f23ecf3355a5bef9/c10/cuda/CUDACachingAllocator.cpp#L131) to make `pool_id` the most significant bit of the search-comparison used when finding a block suitable for a particular stream and pool id.

I don't like this idea as much as lists of per-pool-id Allocators. `BlockComparator` is used in many [BlockPool](https://github.com/pytorch/pytorch/blob/a347c747df8302acc0007a26f23ecf3355a5bef9/c10/cuda/CUDACachingAllocator.cpp#L102) (`std::set`) lookups. Each lookup is logN in the number of blocks the pool contains, so several smaller, distinct Allocators with distinct `BlockPool`s seems a bit better for performance than one Allocator with all blocks from all streams and pool ids in the same `BlockPool`.

#### 4. cudaMallocAsync (implementation alternative)

Cuda 11.2 released a built-in fast asynchronous allocator ([cudaMallocAsync](https://developer.nvidia.com/blog/enhancing-memory-allocation-with-new-cuda-11-2-features/)). It's not capture-safe yet, and some unrelated feature gaps prevent its immediate integration, but it should be upstreamable by 11.3 and capture-safe out of the box _without any manual per-region pools_.

Therefore, private memory pools are in some sense throwaway work because once cudaMallocAsync has been upstreamed (I expect we'll add it as an alternative backend to the current allocator), memory allocations in any region will be safe to capture, as long as they're are satisfied by cudaMallocAsync under the hood.

However, it's not throwaway work if people want to choose between the cudaMallocAsync and current allocator backends in the future, and want to use graphs with both. They'll still need private pools to use the current allocator.

## Additional context

https://github.com/pytorch/pytorch/pull/48875 added primitive graph capture and replay bindings.

I'll post a strawman PR implementing the "Pitch". Writing it is not the hard part.

**Comment by ezyang** [comment]
How will individual allocation requests know which pool they allocate from?  Is it going to be a TLS? (cc @swolchok)

**Comment by mcarilli** [comment]
That's a good question. I would say yes, the context manager writes the MemPool's id to some TLS in CUDACachingAllocator.cpp, but we also need that id to propagate to backward thread(s) when capture is underway, so I'd have to modify the points where backward threads are forked. The same would need to occur for any forward threads, for example, DataParallel would need to retrieve the parent thread's pool id and set it in side threads at the point where its threads fork (it already does so for the autocast state).

**Comment by ezyang** [comment]
Well, all of CUDA caching allocator is under a big honking lock, so it wouldn't be that difficult to safely implement a global flag. Since this is specifically about graph capture, it would make sense to me to have its modality match graph capture; e.g., if graph capture is thread local, then the pooling mechanism should be thread local too. One reason I'm trying to avoid thread local is we have some performance benchmarks that show thread local access is expensive and we are trying to minimize it if not necesary.

**Comment by mcarilli** [comment]
> if graph capture is thread local

Graph capture is [initiated on a Cuda stream](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g793d7d4e474388ddfda531603dc34aa3).  The capture state propagates down the stream and any streams that sync on it. Cuda calls are free-floating from threads, and any threads may enqueue those calls.  In that sense graph capture isn't thread-local, but it isn't entirely global either: unrelated streams running in parallel with capturing streams won't be captured, even if they're enqueued from the same thread that initiated capture.

But our implementation of RNG capture safety does rely on a global state (it sets behavior on the current device's default generator). And we already require only one capture can be underway at a time. And the allocator is already under a lock. Alongside these things, a global flag requesting a private pool seems more natural. My only concerns about the global flag would be, behavior of
```
pool = torch.cuda.MemPool
with torch.cuda.mempool(pool): # toggles a global state
```
isn't symmetric with, for example
```
stream = torch.cuda.Stream()
with torch.cuda.stream(stream): # toggles a thread-local stream state
```
Also, if the private mempool is global, if you want to request a thread local mempool, you can never get one. For graphs, I guess we don't need a thread local mempool, but maybe someone would want one for some unforeseen use case unrelated to graphs.

Since the only use case we have in mind right now is graphs, maybe it's more natural to make the association of graphs and mempools explicit in the API, ie, forget the `with torch.cuda.mempool(pool):` context, and instead have something like
```python
# If you have several graphs and you're sure you'll replay them in a particular sequence, they can share a pool:
pool = getInterGraphMempool()
graph1.capture_begin(pool)
ops...
graph1.capture_end()
graph2.capture_begin(pool)
ops...
graph2.capture_end()

graph1.replay()
graph2.replay()

# If you have several graphs and you want them replayable in any order,
# or replayable simultaneously in different streams:
graph3.capture_begin()# if left to None, the capture will create its own fully-private mempool, invisible to other graphs
ops...
graph3.capture_end()
graph4.capture_begin()
ops...
graph4.capture_end()
with torch.cuda.stream(stream0):
    graph3.replay()
with torch.cuda.stream(stream1):
    graph4.replay()
```

My biggest concern is the entire ask. Is it reasonable? Do you have ideas about a better way to make graphs' baked-in memory safe on replay with the native (non cudaMallocAsync) allocator?

**Comment by ezyang** [comment]
> But our implementation of RNG capture safety does rely on a global state (it sets behavior on the current device's default generator). And we already require only one capture can be underway at a time. And the allocator is already under a lock. Alongside these things, a global flag requesting a private pool seems more natural.

The important distinction here is whether or not this is a temporary state of affairs, or if we want to fundamentally bake this in. Because the underlying CUDA Graph API is stream-specific, and streams are a thread local concept, I'd tend to think that the fact that the PyTorch implementation only has a single global flag is simply an accident of how we are choosing to ship, as opposed to a fundamental implementation. But then that would imply that the allocator changes should operate based on TLS (and in particular, key off of the current stream).

This isn't actually all that strange. The CUDA caching allocator associates allocations with the current stream when you allocated it, see https://github.com/pytorch/pytorch/issues/16668 So I wonder if we're actually closer to where we need to be than you might think.

**Comment by mcarilli** [comment]
> The CUDA caching allocator associates allocations with the current stream when you allocated it,

Aye, **Alternatives 2** and **3** above consider leaning on existing stream allocation association to give private pools.

In Cuda, streams are not a thread local concept, they're free-floating from threads. Any thread may safely enqueue a call to any stream at any time. In Pytorch, the current cuda stream is thread local, but that's an implementation detail of the Pytorch stream pool.

I could imagine the caching allocator checking `currentStreamCaptureStatus()` every time it makes an allocation, and allocating from the current user-specified private pool if so.  This would allow noncapturing streams to use the allocator at the same time as capturing streams, with all requests satisfied from an appropriate pool. In fact, if the capturing streams are guaranteed unique to the graph, we effectively get a "private pool" without needing to check `currentStreamCaptureStatus()` (**Alternative 2**).

As far as baking in user constraints goes, I think
1. only one capture may be underway at a time
2. no unrelated noncaptured work may run while the capture is underway
are reasonable and widen our implementation options. Do you think either of those is too restrictive or would kill some critical use case? (@ngimel @arslan-zulfiqar I'd really like to hear your opinions here too).

**Comment by mcarilli** [comment]
I thought of another reason we may want to associate mempools with graph capture tightly/implicitly. Graphs benefit from a private mempool because if they use the same mempool as everything else, on replay their baked-in memory traffic might corrupt data assigned to more recently-eagerly-allocated tensors. If we capture the graph in a private pool context (for example), there's no danger of replay corrupting eager allocations not created from the private pool.  However, the same danger will resurface if the user eagerly-allocates some tensors from the private pool.  In other words, if a private mempool is used to capture graphs, that pool should _only_ be used to capture graphs.  In other other words, from a graph-safety perspective, having an orthogonal API to set a private-mempool context buys people no more flexibility than a capture API that tightly/implicitly associates a private mempool with the capture itself.

We could make tight/implicit mempool association less verbose by letting each graph object return a handle to its mempool (likely just an integer id) if the user knows it's safe for a sequence of graphs to share the same pool:
```python
graph1.capture_begin() # implicitly grabs a private mempool
graph1.capture_end()

graph2.capture_begin(graph1.mempool()) # tells allocations during graph2's capture to share graph1's mempool
graph2.capture_end()
```

I'm not dogmatic about any approach I've posted so far. And maybe API discussions are cart before horse since it'll be a while before graphs are user-facing. im not even sure private mempools are the best way to ensure allocator safety. This is a tricky problem, made more annoying because cudaMallocAsync will eventually save the day.

**Comment by mcarilli** [comment]
Actually, I just had a reasonable idea how to implement per-graph pools in a way that kind of respects the fact that capture follows streams, as opposed to some strictly global or strictly thread-local "use private pool" flag. The implementation should be simple, too. I'll PR it for discussion.

**Comment by vadimkantorov** [comment]
Would this also allow arena-like allocations/deallocations?

**Comment by mcarilli** [comment]
@vadimkantorov I'm not sure what you mean. For https://github.com/pytorch/pytorch/pull/51436 we decided the simplest way to make the native allocator graph-safe was to make private pools an implementation detail of graph objects, rather than exposing private pools as a separate abstraction.

General Pytorch cudaMallocAsync support will be added later, likely with Cuda 11.3 (the cudaMallocAsync API in 11.2 lacks some crucial features pytorch needs). Most likely, we'll keep the current allocator as an alternative backend, so people can choose to use cudaMallocAsync or a native open-source allocator.

**Comment by ezyang** [comment]
@vadimkantorov Actually, technically, I think this would make it possible. 

@mcarilli I believe what Vadim is looking for is a way to do a bunch of CUDA allocations *without* freeing any of them, and then freeing the entire chunk of allocations all in one go at the end. (Actually, I'm not really sure this is a good idea for typical PyTorch code uses). But it seems to me like private memory pools function similarly, since you are doing allocations off into a private pool; in the case of CUDA graphs it is to make sure you keep that memory around for future CUDA graph allocations, but it could easily also just be to let you free it all in one go later.

**Comment by vadimkantorov** [comment]
> and then freeing the entire chunk of allocations all in one go at the end. (

That's one thing

Another thing:
Completely crush fragmentation after an iteration.
Preallocate a workspace
Potentially configure an allocator (trace allocations, pre-specify allocation size bins)

Often because of unlucky tensor sizes, we get OOM on subsequent iterations of training.

**Comment by ezyang** [comment]
Updating the state on this: CUDA graphs have private memory pools, but we didn't expose a public API for people to generally use this, so I'm keeping this issue open

**Comment by FuncJ** [comment]
> > if graph capture is thread local
> 
> Graph capture is [initiated on a Cuda stream](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g793d7d4e474388ddfda531603dc34aa3). The capture state propagates down the stream and any streams that sync on it. Cuda calls are free-floating from threads, and any threads may enqueue those calls. In that sense graph capture isn't thread-local, but it isn't entirely global either: unrelated streams running in parallel with capturing streams won't be captured, even if they're enqueued from the same thread that initiated capture.
> 
> But our implementation of RNG capture safety does rely on a global state (it sets behavior on the current device's default generator). And we already require only one capture can be underway at a time. And the allocator is already under a lock. Alongside these things, a global flag requesting a private pool seems more natural. My only concerns about the global flag would be, behavior of
> 
> ```
> pool = torch.cuda.MemPool
> with torch.cuda.mempool(pool): # toggles a global state
> ```
> 
> isn't symmetric with, for example
> 
> ```
> stream = torch.cuda.Stream()
> with torch.cuda.stream(stream): # toggles a thread-local stream state
> ```
> 
> Also, if the private mempool is global, if you want to request a thread local mempool, you can never get one. For graphs, I guess we don't need a thread local mempool, but maybe someone would want one for some unforeseen use case unrelated to graphs.
> 
> Since the only use case we have in mind right now is graphs, maybe it's more natural to make the association of graphs and mempools explicit in the API, ie, forget the `with torch.cuda.mempool(pool):` context, and instead have something like
> 
> ```python
> # If you have several graphs and you're sure you'll replay them in a particular sequence,
> # they can share a pool:
> pool = getInterGraphMempool()
> graph1.capture_begin(pool)
> ops...
> graph1.capture_end()
> graph2.capture_begin(pool)
> ops...
> graph2.capture_end()
> 
> graph1.replay()
> graph2.replay()
> 
> # If you have several graphs and you want them replayable simultaneously,
> # they may not share a pool:
> graph3.capture_begin()# if left to None, the capture will create its own fully-private mempool, invisible to other graphs
> ops...
> graph3.capture_end()
> graph4.capture_begin()
> ops...
> graph4.capture_end()
> with torch.cuda.stream(stream0):
>     graph3.replay()
> with torch.cuda.stream(stream1):
>     graph4.replay()
> ```
> 
> My biggest concern is the entire ask. Is it reasonable? Do you have ideas about a better way to make graphs' baked-in memory safe on replay with the native (non cudaMallocAsync) allocator?

Hello, could you please explain the principle behind different graphs reusing memory space? Could you spare some time to help me resolve this confusion? The link is as follows: https://github.com/pytorch/pytorch/issues/137844#issue-2582738292
